### PR TITLE
Arm vhe nested virt

### PR DESCRIPTION
**Pull request by ardbiesheuvel**
Virtual machines generally only execute at EL1, and even if ArmVirtQemu happens to work when executing at EL2, it was never the intended mode of operation

With nested virtualization, the situation is a bit different. Virtual machines can be expected to run at EL2 even in production cases using real hardware, and the architecture now permits VHE (virtualization host extensions) to be the only implemented execution mode at EL2.

This series contains the necessary changes to deal with the above, and to make the EDK2 build for the ARM KvmTool platform compatible with execution at EL2.

The first patch has already been sent out separately but it is included here for completeness. The resulting build can boot a stock Debian install (with 16k pages kernel) successfully on Apple M2 silicon in a VM running at EL2, and host nested VMs using the distro provided QEMU and ArmVirtQemu packages.

### CODE REVIEW

**Review by os-d — APPROVED** [review]

**Comment by jrtc27** [comment]
> Virtual machines generally only execute at EL1, and even if ArmVirtQemu happens to work when executing at EL2, it was never the intended mode of operation

This is, however, not so unusual when using QEMU for system emulation rather than KVM, with just `-M virtualization=on` needed to enable it. It's slow, but a common way to test or debug a hypervisor (and what I was trying to do that discovered the CPTR_EL2 RES1 issue).

**Comment by ardbiesheuvel** [comment]
@samimujawar any comments on this series?

File: ArmPkg/Library/ArmLib/AArch64/AArch64ArchTimerSupport.S [hunk 1]
```diff
@@ -9,6 +9,41 @@
...
 
+/* Special accessors for VHE */
+#define cntp_tval_el02          s3_5_c14_c2_0
+#define cntp_ctl_el02           s3_5_c14_c2_1
+#define cntp_cval_el02          s3_5_c14_c2_2
+#define cntv_tval_el02          s3_5_c14_c3_0
+#define cntv_ctl_el02           s3_5_c14_c3_1
+#define cntv_cval_el02          s3_5_c14_c3_2
+
+#define RVHE(reg, sysreg)       rvhe reg, sysreg, sysreg ## 2
```
Comment (samimujawar) [inline ArmPkg/Library/ArmLib/AArch64/AArch64ArchTimerSupport.S:20]: Minor, should we rename the macro names to RCNT as in read counter and  WCNT write counter. The only reason being the macros internally make the decision to call the VHE specific version based on the HCR_EL2.E2H bit value.

File: ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibCore.c [hunk 2]
```diff
@@ -568,6 +568,8 @@ ArmConfigureMmu (
   UINT64      TCR;
   EFI_STATUS  Status;
 
+  ASSERT (ArmReadCurrentEL () < AARCH64_EL3);
+
```
Comment (samimujawar) [inline ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibCore.c:589]: Should we also return EFI_UNSUPPORTED? The existing code had this behaviour, however I acknowledge that the return value is not being checked.

**Review by samimujawar — APPROVED** [review]
I have some minor review comments, other than that this patch series looks good to me.

File: ArmPkg/Library/ArmLib/AArch64/AArch64ArchTimerSupport.S (hunk 1 above)
Comment (ardbiesheuvel) [reply ArmPkg/Library/ArmLib/AArch64/AArch64ArchTimerSupport.S:20]: Fair enough

File: ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibCore.c (hunk 2 above)
Comment (ardbiesheuvel) [reply ArmPkg/Library/ArmMmuLib/AArch64/ArmMmuLibCore.c:589]: Yes that makes sense.

**Pull request merged**